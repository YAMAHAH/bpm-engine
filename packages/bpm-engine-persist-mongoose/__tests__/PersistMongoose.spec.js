import BPMEngine from 'bpm-engine';
import fs from 'fs';
import Promise from 'bluebird';

import mongoose from 'mongoose';

import PersistMongoose from 'bpm-engine-persist-mongoose';

import History from '../../bpm-engine/__tests__/Plugins/History';
import sleep from '../../bpm-engine/__tests__/Plugins/sleep';

mongoose.Promise = Promise;

describe('PersistMongoose', () => {
  let persistMongoose;

  beforeEach(async () => {
    persistMongoose = new PersistMongoose('mongodb://localhost:27017/bpm-engine-testing', {
      useNewUrlParser: true,
      useFindAndModify: false,
      useCreateIndex: true,
    });

    // clear collections
    const clearCollections = [];
    clearCollections.push(persistMongoose.schemas.processInstance.deleteMany({}).exec());
    clearCollections.push(persistMongoose.schemas.tokenInstance.deleteMany({}).exec());
    clearCollections.push(persistMongoose.schemas.workflowDefinition.deleteMany({}).exec());
    clearCollections.push(persistMongoose.schemas.timer.deleteMany({}).exec());
    clearCollections.push(persistMongoose.schemas.task.deleteMany({}).exec());

    await Promise.all(clearCollections);
  });

  afterEach(async () => {
    await persistMongoose.connection.close();
  });

  it('Creates a process instance by xml', async () => {
    const history = new History();

    const engine = new BPMEngine({
      plugins: [history],
      persist: persistMongoose,
    });

    const processInstance = await engine.createProcessInstance({
      workflowDefinition: fs.readFileSync(`${__dirname}/ParallelServices.bpmn`, 'utf-8'),
    });

    await processInstance.execute();

    expect((await persistMongoose.schemas.processInstance.find()).length).toMatchSnapshot();
    expect((await persistMongoose.schemas.tokenInstance.find()).length).toMatchSnapshot();
    expect(history.store).toMatchSnapshot();
    engine.clock.stop();
  });

  it('Can deploy a workflow definition', async () => {
    const engine = new BPMEngine({
      persist: persistMongoose,
    });

    await engine.deployWorkflowDefinition({
      xml: fs.readFileSync(`${__dirname}/ParallelServices.bpmn`, 'utf-8'),
    });

    expect((await persistMongoose.schemas.workflowDefinition.find()).length).toMatchSnapshot();
    engine.clock.stop();
  });

  it('Can create a process instance from a deployed workflowDefinition', async () => {
    const engine = new BPMEngine({
      persist: persistMongoose,
    });

    const deployedWorkflowDefinition = await engine.deployWorkflowDefinition({
      xml: fs.readFileSync(`${__dirname}/ParallelServices.bpmn`, 'utf-8'),
    });

    const processInstance = await engine.createProcessInstance({
      workflowDefinitionId: deployedWorkflowDefinition.workflowDefinitionId,
    });

    expect(processInstance).not.toBeFalsy();
    engine.clock.stop();
  });

  it('Can create a process instance with a timer start event', async () => {
    const bpm = new BPMEngine({
      persist: persistMongoose,
    });

    await bpm.deployWorkflowDefinition({
      xml: fs.readFileSync(`${__dirname}/TimerStartEvent.bpmn`, 'utf-8'),
    });

    await sleep(3000);
    const results = await persistMongoose.schemas.processInstance.find();
    expect(results.length).toBe(3);
    bpm.clock.stop();
  });

  it('multiple deploys of the same workfowDefinition should replace existing timers', async () => {
    const engine = new BPMEngine({
      persist: persistMongoose,
    });

    await engine.deployWorkflowDefinition({
      xml: fs.readFileSync(`${__dirname}/TimerStartEvent.bpmn`, 'utf-8'),
      workflowDefinitionId: 'timerStartEvent',
    });

    await sleep(1500);
    let timers = await persistMongoose.schemas.timer.find();
    let lastTimer = timers[2];

    expect(lastTimer.status).toBeFalsy();

    await engine.deployWorkflowDefinition({
      xml: fs.readFileSync(`${__dirname}/TimerStartEvent.bpmn`, 'utf-8'),
      workflowDefinitionId: 'timerStartEvent',
    });

    timers = await persistMongoose.schemas.timer.find();
    lastTimer = timers[2];
    expect(lastTimer.status).toEqual('done');

    await sleep(1500);
    timers = await persistMongoose.schemas.timer.find();
    lastTimer = timers[5];
    expect(lastTimer.status).toBeFalsy();
    engine.clock.stop();
  });
});
