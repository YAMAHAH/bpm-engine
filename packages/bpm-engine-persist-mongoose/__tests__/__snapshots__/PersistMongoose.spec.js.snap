// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PersistMongoose Can deploy a workflow definition 1`] = `1`;

exports[`PersistMongoose Creates a process instance by xml 1`] = `1`;

exports[`PersistMongoose Creates a process instance by xml 2`] = `3`;

exports[`PersistMongoose Creates a process instance by xml 3`] = `
Array [
  Object {
    "elementId": "StartEvent_1",
    "lanes": undefined,
    "participant": undefined,
    "state": "ready",
  },
  Object {
    "elementId": "StartEvent_1",
    "lanes": undefined,
    "participant": undefined,
    "state": "active",
  },
  Object {
    "elementId": "StartEvent_1",
    "lanes": undefined,
    "participant": undefined,
    "state": "complete",
  },
  Object {
    "elementId": "Task_0rnu0cc",
    "lanes": undefined,
    "participant": undefined,
    "state": "ready",
  },
  Object {
    "elementId": "Task_0rnu0cc",
    "lanes": undefined,
    "participant": undefined,
    "state": "active",
  },
  Object {
    "elementId": "Task_0rnu0cc",
    "lanes": undefined,
    "participant": undefined,
    "state": "complete",
  },
  Object {
    "elementId": "ExclusiveGateway_03c6fwo",
    "lanes": undefined,
    "participant": undefined,
    "state": "ready",
  },
  Object {
    "elementId": "ExclusiveGateway_03c6fwo",
    "lanes": undefined,
    "participant": undefined,
    "state": "active",
  },
  Object {
    "elementId": "ExclusiveGateway_03c6fwo",
    "lanes": undefined,
    "participant": undefined,
    "state": "complete",
  },
  Object {
    "elementId": "Task_0z13d9i",
    "lanes": undefined,
    "participant": undefined,
    "state": "ready",
  },
  Object {
    "elementId": "Task_0z13d9i",
    "lanes": undefined,
    "participant": undefined,
    "state": "active",
  },
  Object {
    "elementId": "Task_0z13d9i",
    "lanes": undefined,
    "participant": undefined,
    "state": "complete",
  },
  Object {
    "elementId": "ExclusiveGateway_0jwxcj8",
    "lanes": undefined,
    "participant": undefined,
    "state": "ready",
  },
  Object {
    "elementId": "Task_1ro3bn5",
    "lanes": undefined,
    "participant": undefined,
    "state": "ready",
  },
  Object {
    "elementId": "Task_1ro3bn5",
    "lanes": undefined,
    "participant": undefined,
    "state": "active",
  },
  Object {
    "elementId": "Task_1ro3bn5",
    "lanes": undefined,
    "participant": undefined,
    "state": "complete",
  },
  Object {
    "elementId": "ExclusiveGateway_0jwxcj8",
    "lanes": undefined,
    "participant": undefined,
    "state": "ready",
  },
  Object {
    "elementId": "ExclusiveGateway_0jwxcj8",
    "lanes": undefined,
    "participant": undefined,
    "state": "active",
  },
  Object {
    "elementId": "ExclusiveGateway_0jwxcj8",
    "lanes": undefined,
    "participant": undefined,
    "state": "complete",
  },
  Object {
    "elementId": "Task_0ihlvap",
    "lanes": undefined,
    "participant": undefined,
    "state": "ready",
  },
  Object {
    "elementId": "Task_0ihlvap",
    "lanes": undefined,
    "participant": undefined,
    "state": "active",
  },
  Object {
    "elementId": "Task_0ihlvap",
    "lanes": undefined,
    "participant": undefined,
    "state": "complete",
  },
  Object {
    "elementId": "EndEvent_15k76o1",
    "lanes": undefined,
    "participant": undefined,
    "state": "ready",
  },
  Object {
    "elementId": "EndEvent_15k76o1",
    "lanes": undefined,
    "participant": undefined,
    "state": "active",
  },
  Object {
    "elementId": "EndEvent_15k76o1",
    "lanes": undefined,
    "participant": undefined,
    "state": "complete",
  },
]
`;
